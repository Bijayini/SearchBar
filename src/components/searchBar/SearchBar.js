import React, { useEffect, useState, useRef } from 'react';

import { URL } from '../../constants';
import {debounce, isStringPresentInElement} from "../../utils/helper";
import SearchList from "../seachList/SearchList";
import WithOutsideClick from "../outSideClick/OutSideClick";

import('./searchBar.scss');

const SearchBar = () => {
  const [users, setUsers] = useState([]);
  const [lists, setLists] = useState([]);
  const [showList, setShowList] = useState(false);
  const [pointer, setPointer] = useState(-1);
  const [inputValue, setInputValue] = useState('');
  const activeItem = useRef(null);


  useEffect(() => {
    fetch(URL)
      .then(results => results.json())
      .then(data => setUsers(data))
      .catch(error => console.log(error));
  }, []);


  const filterUser = input => {
    return users.filter((user) => {
      const keys = Object.keys(user);
      return keys.some(key => {
        if (typeof user[key] !== 'object') {
          return isStringPresentInElement(user[key], input);
        }
        return  (user[key].some(item =>isStringPresentInElement(item, input)));
      });
    });
  };

  const ensureActiveItemVisible =()=> {
    if (activeItem.current) {
      activeItem.current.scrollIntoView({
        behavior: 'smooth'
      });
    }
    console.log('after updating', pointer, activeItem);
  };

  const searchAndShowList = (value) =>{
    setShowList(true);
    setLists(filterUser(value));
  };

  const search = debounce(searchAndShowList, 300);

  const handleChange = (value) => {
    if(value){
      setInputValue(value);
      search(value);
    }else if(value === '') {
      setInputValue(value);
      setShowList(false);
      setLists([]);
    }
  };

  const updatePointer = e => {
    if(e.keyCode === 40){
      if(pointer < lists.length-1){
        setPointer(pointer + 1);
      }else if(pointer === lists.length - 1){
        setPointer(-1);
      }
    }else if(e.keyCode === 38){
      if(pointer > -1){
        setPointer(pointer - 1);
      }else if(pointer === -1){
        setPointer(0);
      }
    }
  };
  return (
    <WithOutsideClick
      closeDropdown={() => {
        setShowList(false)
      }}
    >
      <div className="field-container" aria-haspopup="listbox"
           role="combobox"
           aria-owns="suggestionContainer"
           aria-expanded={showList}
           id="autoSuggestWrapper"
      >
        <input
          type="text"
          className="search-input"
          dataTestId="searchInput"
          placeholder="Search User By Id, address, name,items, pincode"
          onChange={(e)=>handleChange(e.target.value)}
          onKeyDown={updatePointer}
          onKeyUp={ensureActiveItemVisible}
          value={inputValue}
        />
        {inputValue && <button onClick={()=>handleChange('')} className="close-icon">&#10005;</button>}
        {showList && (
          <SearchList setPointer={setPointer} pointer={pointer} lists={lists} input={inputValue} activeItem={activeItem} />
        )}
      </div>
    </WithOutsideClick>
  );
};

export default SearchBar;
