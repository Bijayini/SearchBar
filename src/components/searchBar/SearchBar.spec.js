import React from "react";
import { render, act, fireEvent } from "@testing-library/react";
import SearchBar from "./SearchBar";

let component;
global.fetch = jest.fn(() => Promise.resolve({ json: () => [{
    "id": "123-s2-546",
    "name": "John Jacobs",
    "items": ["bucket", "bottle"],
    "address": "1st Cross, 9th Main, abc Apartment",
    "pincode": "5xx012"
  }]
}));

beforeEach(async() => {
  await act(async () => {
    component = render(<SearchBar />).container;
  });
});

describe("SearchBar component", () => {
  it("should render SearchBar without lists", () => {
    expect(component).toMatchSnapshot();
  });
});
