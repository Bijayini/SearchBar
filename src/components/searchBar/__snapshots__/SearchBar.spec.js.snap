// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component should render SearchBar without lists 1`] = `
<div>
  <div
    class="overlay"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="suggestionContainer"
      class="field-container"
      id="autoSuggestWrapper"
      role="combobox"
    >
      <input
        class="search-input"
        datatestid="searchInput"
        placeholder="Search User By Id, address, name,items, pincode"
        type="text"
        value=""
      />
      
    </div>
  </div>
</div>
`;
